File: BaseLibrary/communication/protocol/Factories.cs
using System.Collections.Generic;

namespace BaseLibrary.communication.protocol {
    public class Factories<TResult, TTransfer> {
        private readonly List<IFactory<TResult, TTransfer>> commandFactories = new List<IFactory<TResult, TTransfer>>();

        public void RegisterCommand(IFactory<TResult, TTransfer> factory) {
            if (factory == null) return;
            commandFactories.Add(factory);
        }

        public void UnregisterCommand(IFactory<TResult, TTransfer> factory) {
            if (factory == null) return;
            commandFactories.Remove(factory);
        }

        public TResult Deserialize(string transferString) {
            if (transferString != null) {
                foreach (IFactory<TResult, TTransfer> commandFactory in commandFactories) {
                    if (commandFactory.IsDeserializeable(transferString)) {
                        return commandFactory.Deserialize(transferString);
                    }
                }
            }
            return default(TResult);
        }

        public TResult Transfer(TTransfer transfer) {
            if (transfer != null) {
                foreach (IFactory<TResult, TTransfer> commandFactory in commandFactories) {
                    if (commandFactory.IsTransferable(transfer)) {
                        return commandFactory.Transfer(transfer);
                    }
                }
            }
            return default(TResult);
        }

        public string Serialize(TTransfer transfer) {
            if (transfer != null) {
                foreach (IFactory<TResult, TTransfer> commandFactory in commandFactories) {
                    if (commandFactory.IsSerializable(transfer)) {
                        return commandFactory.Serialize(transfer);
                    }
                }
            }
            return default(string);
        }
    }
}

File: Java-Client/RobotBattlefield-Client-Java/src/BaseLibrary/communication/protocol/Factories.java
package BaseLibrary.communication.protocol;

import java.util.ArrayList;
import java.util.List;

public class Factories<TResult, TTransfer> {
	private final List<IFactory<TResult, TTransfer>> commandFactorys = new ArrayList<IFactory<TResult, TTransfer>>();

	public void RegisterCommand(IFactory<TResult, TTransfer> factory) {
		commandFactorys.add(factory);
	}

	public void UnregisterCommand(IFactory<TResult, TTransfer> factory) {
		commandFactorys.remove(factory);
	}

	public TResult Deserialize(String transferString, AProtocol protocol) {
		for (IFactory<TResult, TTransfer> commandFactory : commandFactorys) {
			if (commandFactory.isDeserializeable(transferString, protocol)) {
				return commandFactory.deserialize(transferString, protocol);
			}
		}
		return null;
	}

	public String Serialize(TTransfer transfer, AProtocol protocol) {
		for (IFactory<TResult, TTransfer> commandFactory : commandFactorys) {
			if (commandFactory.isSerializeable(transfer, protocol)) {
				return commandFactory.serialize(transfer, protocol);
			}
		}
		return null;
	}
}
